articles/zero_copy.md

```diff
@@ -0,0 +1,249 @@
+The user should be able to influence what the `Executor` does, and in the case that memory needs to be allocated, the user should be able to provide an allocator or memory management strategy which would influence the `Executor`'s behavior.
+
+## Requirements
+
+Based on the the use cases above, the general requirements are as follows:
+
+* Users must be able to avoid all memory operations and copies in at least one configuration.
+
+
+### Publisher Requirements
+
+* The user must be able to publish from messages allocated in their stack or heap.
+* The user must be able to get a loaned message, use it, and return it during publication.
+* When taking a loan, the middleware should not do anything "special", that is that the user must be able to influence the allocation.
+
+### Subscription Requirements
+
+The following requirements should hold whether the user is polling or using a wait set.
+
+* The user must be able to have the middleware fill data into messages allocated in the user's stack or heap.
+* The user must be able to get a loaned message from the middleware when calling take.
+* The user must be able to get a sequence of loaned messages from the middleware when calling take.
+* The loaned message or sequence must be returned by the user.
+* When taking a loan, the middleware should not do anything "special", that is that the user must be able to influence the allocation.
+
+### Special Requirements
+
+These requirements are driven by idiosyncrasies of various middleware implementations and some of their special operating modes, e.g. zero copy:
+
+#### RTI Connext DDS
+
+* The Connext API (more generally the DDS API) requires that the user to use a sequence of messages when taking or reading.
+  * This means the ROS API needs to do the same, otherwise the middleware would be giving a "loan" to a message in a sequence, but it would also need to keep the sequence immutable.
+* Needs to keep sample and sample info together, therefore `rclcpp` loaned message sequence needs to own both somehow.
+
+Connext Micro Specific (ZeroCopy):
+
+* The user needs to be able to check if data has been invalidated after reading it.
+
+### Nice to Haves
+
+* When using loaned messages, the overhead should be minimized, so that it can be used as the general approach as often as possible.
+  * Versus reusing a user owned message on the stack or heap.
+  * We will still recommend reusing a message repeatedly from the stack or heap of the user.
+* Memory operations and copies should be avoided anywhere possible.
+
+## Design Proposal
+
+### RMW API
+
+Introduce APIs for creating/destroying loaned messages, as well as structure for management:
+
+```
+void *
+rmw_allocate_loaned_message(
+  const rmw_publisher_t * publisher,
+  const rosidl_message_type_support_t * type_support,
+  size_t message_size
```

Comment:
While some messages are PODs many messages containing sequences are not. In the pending PR in `rclcpp` e.g. the value `sizeof(MessageT)` is passed for this argument. How does that handle the memory allocation of the sequence fields since it doesn't seem to be the case that the RMW interface is aware of how many items should be allocated for the message?

Reply:
When creating a publisher, it has to be distinguished whether the publisher can loan messages of the given message type or not. If not, the call to `rmw_allocate_loaned_message` shall fail.

Reply:
Your response doesn't rely answer my question. I understand that if loaning isn't supported the calling code can detect that and use a different approach if desired. So the question remains: is this design draft and the implementation limited to POD message? If yes, this needs to be mentioned in this article imo. Also from my point of few that would be a severe limitation of the proposal and would need a very good rational why non-POD aren't / can't be supported (since they are the majority of message types).

Reply:
non-POD must be supported. The only thing not being guaranteed with non-POD is that a zero-copy mechanism can be supported. That is, for heap allocated data structures, copies and eventual serializations have to be done. 
But I see your point that this has to be noted more explicitly in this design doc.

Reply:
My concern is that I am not convinced that the currently proposed API is capable of supporting non-POD messages. And if that is the case I would like to see an API which is able to support that or a rational why that is not possible / feasible.

Reply:
> My concern is that I am not convinced that the currently proposed API is capable of supporting non-POD messages. And if that is the case I would like to see an API which is able to support that or a rational why that is not possible / feasible.

Can you please comment on my last reply when you have a chance.

Reply:
The API in question has been removed in the mean time. The new signature of `rmw_borrow_loaned_message` is not limited like this so this looks fine to me.